reference-implementation/lib/readable-stream.js

```diff
@@ -156,67 +131,349 @@ export default class ReadableStream {
     }
 
     function abortDest(reason) {
-      source.releaseLock();
+      // Does not need to wait for lastRead since it only occurs on source errored.
+
+      reader.releaseLock();
+
       if (preventAbort === false) {
         dest.abort(reason);
       }
       rejectPipeToPromise(reason);
     }
   }
+}
+
+class ReadableStreamReader {
+  constructor(stream) {
+    if (IsReadableStream(stream) === false) {
+      throw new TypeError('ReadableStreamReader can only be constructed with a ReadableStream instance');
+    }
+    if (IsReadableStreamLocked(stream) === true) {
+      throw new TypeError('This stream has already been locked for exclusive reading by another reader');
+    }
+
+    stream._reader = this;
+    this._ownerReadableStream = stream;
+    this._state = 'readable';
+    this._storedError = undefined;
+
+    this._readRequests = [];
+
+    this._closedPromise = new Promise((resolve, reject) => {
+      this._closedPromise_resolve = resolve;
+      this._closedPromise_reject = reject;
+    });
+
+    if (stream._state === 'closed' || stream._state === 'errored') {
+      ReleaseReadableStreamReader(this);
+    }
+  }
+
+  get closed() {
+    if (IsReadableStreamReader(this) === false) {
+      return Promise.reject(
+        new TypeError('ReadableStreamReader.prototype.closed can only be used on a ReadableStreamReader'));
+    }
+
+    return this._closedPromise;
+  }
+
+  cancel(reason) {
+    if (IsReadableStreamReader(this) === false) {
+      return Promise.reject(
+        new TypeError('ReadableStreamReader.prototype.cancel can only be used on a ReadableStreamReader'));
+    }
+
+    if (this._ownerReadableStream === undefined) {
+      return Promise.resolve(undefined);
+    }
```

Comment:
Return a promise rejected with the error if this was released when it got errored?